integr8ly/tutorial-web-app
# APPDUX-167: First phase of configuration management

## Motivation
**JIRA:**
https://issues.redhat.com/browse/APPDUX-167

**Design spec:**
https://marvelapp.com/11id0249/screen/69959965

## What
Add multi-tab interface for Solution Explorer settings page. Revise existing solution patterns tab and create new Managed Integration schedule-related tab as described by JIRA story and design spec. The new tab:

- Fetches the daily backup and maintenance window data from the rhmi config, and formats as to spec in local time and UTC time
- Allows you to write a new start time for daily backups
- Prevents daily backups from being scheduled during first hour of next maintenance window (by disabling it in the dropdown)

## Why
Allow users to make backend OpenShift changes directly via Solution Explorer vs. editing the config YAML in OpenShift.

## Verification Steps
1. Log in with admin account.
2. Go to the Settings page.
3. Click the Managed Integration schedule tab.
4. Select a start time for backups and click Save.
5. Verify that if you go to settings again, the Next daily backup date/time has changed. Bonus points, go to the RHMIConfigs YAML on OpenShift and verify that the values have changed accordingly.
6. Verify that the Start time dropdown displays the current Next daily backup.
7. Verify that the weekly maintenance window date/time reflects what is in the YAML file.
8. Verify that the backup dropdown has disabled the time in which the weekly maintenance window occurs so that it is not selectable.
9. Click the Solution Pattern content tab.
10. Verify that the syntax checking and ability to add new github repos still works.
11. Verify that the new 'Learn more...' link at the bottom navigates to the expected URL. Note that currently it loads the correct page but it looks like Doc has not created the expected bookmark. Not a bug with this code.

## Checklist:

## Progress

## Additional Notes
Functionality is complete, and most CSS and other related changes are complete. Sending this as a draft pull request because of known CSS issues that are being fixed... wanted to get the core functionality/code into review.

**Known CSS issues:**
- **(FIXED)** Selectable area on tab card should not be selectable
- **(FIXED)** Daily backups dropdown is too wide
- **(FIXED)** Need to remove extra gray space at top where 'last saved' info will eventually be
- **(FIXED)** 'Permissions needed' page should expand to new tab size

**Testing:**
For now, you can test on my OpenShift cluster (available until 6/27):
https://solution-explorer.apps.cluster-uxddev-730d.uxddev-730d.example.opentlc.com/

This build disables security so you can log in as a non-admin, for ease of testing. Can also load a version with security if desired.

Also, you can load either of the following docker images onto your own OpenShift cluster, the first disables security so you don't need to be logged in as Admin, the second has security enabled and is how it will eventually be released:

docker.io/mfrances17/dev-tutorial-web-app:configNoSecurity
docker.io/mfrances17/dev-tutorial-web-app:config

**Screen caps**

**Backup tab:**
![backup_tab](https://user-images.githubusercontent.com/39063664/85596634-e58dcd00-b617-11ea-8605-488e91b8179c.png)

**Solution Patterns tab:**
![solution_pattern_tab](https://user-images.githubusercontent.com/39063664/85596604-deff5580-b617-11ea-87bb-e00f1d0e4083.png)

**Conflicting maintenance window time disabled:**
![maint_time_disabled](https://user-images.githubusercontent.com/39063664/85596612-e161af80-b617-11ea-94ef-3ac890679b5d.png)

**RHMIConfig YAML write succeded:**
![YML_config_changed](https://user-images.githubusercontent.com/39063664/85596580-d9a20b00-b617-11ea-98af-09d10d1a9503.png)
